File: src/main/java/org/droidmate/analyzer/BatchProcessor.java
package org.droidmate.analyzer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

/**
 * Perform experiment for multiple applications
 */
class BatchProcessor {
    private static final Logger logger = LoggerFactory.getLogger(BatchProcessor.class);

    private List<IAppUnderTest> apps;
    private Configuration cfg;
    private AppAnalyzer analyzer;

    BatchProcessor(Configuration cfg) {
        this.cfg = cfg;
        this.analyzer = new AppAnalyzer(cfg);
        this.apps = new ArrayList<>();
    }

    private void initializeApkList() {
        Path dirPath = Paths.get(this.cfg.apkInputDir);
        try {
            Stream<Path> files = Files.list(dirPath);
            files.filter(p -> p.getFileName().toString().contains(".apk")).forEachOrdered(apkPath ->
                    {
                        IAppUnderTest apk = new AppUnderTest(this.cfg, apkPath);
                        this.apps.add(apk);
                        logger.debug(apk.getPackageName());
                    }
            );
        } catch (IOException e) {
            logger.error(e.getMessage(), e);
        }
    }

    void analyze() {
        this.initializeApkList();

        ReportGenerator reporter = new ReportGenerator();

        this.apps.forEach(apk -> {
            logger.info(String.format("Executing app %s", apk.toString()));
            this.analyzer.analyze(apk, reporter);
            reporter.generateReport();
        });
    }
}

File: src/main/kotlin/org/droidmate/analyzer/BatchProcessor.kt
package org.droidmate.analyzer

import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

/**
 * Perform experiment for multiple applications
 */
internal class BatchProcessor(private val cfg: Configuration) {

    private val apps: MutableList<IAppUnderTest>
    private val analyzer: AppAnalyzer = AppAnalyzer(cfg)

    init {
        this.apps = ArrayList<IAppUnderTest>()
    }

    private fun initializeApkList() {
        val dirPath = Paths.get(this.cfg.inputDir)
        try {
            val files = Files.list(dirPath)
            files.filter { p -> p.fileName.toString().contains(".apk") }
                    .forEachOrdered { apkPath ->
                        val apk = AppUnderTest(this.cfg, apkPath)
                        this.apps.add(apk)
                        logger.debug(apk.packageName)
                    }
        } catch (e: IOException) {
            logger.error(e.message, e)
        }
    }

    fun analyze() {
        this.initializeApkList()

        val reporter = ReportGenerator()

        this.apps.forEach { apk ->
            logger.info(String.format("Executing app %s", apk.toString()))
            this.analyzer.analyze(apk, reporter)
            reporter.generateReport()
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(BatchProcessor::class.java)
    }
}
